HeterogeneousCore/SonicTriton/interface/TritonService.h

```diff
@@ -0,0 +1,115 @@
+#ifndef HeterogeneousCore_SonicTriton_TritonService
+#define HeterogeneousCore_SonicTriton_TritonService
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+#include "tbb/concurrent_unordered_set.h"
+
+#include <vector>
+#include <unordered_set>
+#include <string>
+#include <functional>
+#include <utility>
+
+//forward declarations
+namespace edm {
+  class ActivityRegistry;
+  class ConfigurationDescriptions;
+  class PathsAndConsumesOfModulesBase;
+  class ProcessContext;
+}  // namespace edm
+
+class TritonService {
+public:
+  //classes and defs
+  struct FallbackOpts {
+    FallbackOpts(const edm::ParameterSet& pset)
+        : enable(pset.getUntrackedParameter<bool>("enable")),
+          verbose(pset.getUntrackedParameter<bool>("verbose")),
+          useDocker(pset.getUntrackedParameter<bool>("useDocker")),
+          useGPU(pset.getUntrackedParameter<bool>("useGPU")),
+          retries(pset.getUntrackedParameter<int>("retries")),
+          wait(pset.getUntrackedParameter<int>("wait")),
+          instanceName(pset.getUntrackedParameter<std::string>("instanceName")),
+          tempDir(pset.getUntrackedParameter<std::string>("tempDir")) {}
+
+    bool enable;
+    bool verbose;
+    bool useDocker;
+    bool useGPU;
+    int retries;
+    int wait;
+    std::string instanceName;
+    std::string tempDir;
+  };
+  struct Server {
+    Server(const edm::ParameterSet& pset)
+        : name(pset.getUntrackedParameter<std::string>("name")),
+          url(pset.getUntrackedParameter<std::string>("address") + ":" +
+              std::to_string(pset.getUntrackedParameter<unsigned>("port"))),
+          isFallback(name == fallbackName) {}
+    Server(const std::string& name_, const std::string& url_ = "")
+        : name(name_), url(url_), isFallback(name == fallbackName) {}
+
+    struct Hash {
+      size_t operator()(const Server& obj) const { return hashObj(obj.name); }
+      std::hash<std::string> hashObj;
+    };
+
+    struct Equal {
+      bool operator()(const Server& lhs, const Server& rhs) const { return lhs.name == rhs.name; }
+    };
+
+    //members
+    std::string name;
+    std::string url;
+    bool isFallback;
+    mutable std::unordered_set<std::string> models;
```

Comment:
`unordered_set` only provides const access to its elements, because the design assumes the entire object is used to compute the key. However, the `Hash` and `Equal` operators for this struct only consider the "name" field, so the `models` field can be changed without changing the hash value of the object. Therefore, `mutable` indicates it is safe to be modified.

Reply:
Is
https://github.com/cms-sw/cmssw/blob/5c7b0a1053011c4e0686c35160a036987d066456/HeterogeneousCore/SonicTriton/interface/TritonService.h#L111
then really a map from `name` to the rest of the elements of `Server`? How about dropping `name` from `Server` and using `std::unordered_map<std::string, Server>`?

Reply:
Yes, I suppose that's a more concise way to implement it. (It just bugs me that C++ doesn't have a nicer way to represent "name" as a property of `Server` while in a set.)